### ISSUE TITLE
Hyperion Phase 4: agents, the routing prior, and reliability

### ISSUE THREAD

**Issue by dhruuvsharma** [dhruuvsharma/DaxAlgo-Terminal#48]
Phase 4 of #44. Buildable now because the two things it depends on exist: the ladder scores an agent's
work objectively (#46), and registration delivers what it produces.

## The shape

Six agents, each **paired with the rung that scores it**. That pairing is what makes reward-biased
routing meaningful rather than decorative — reliability is measured against an objective check, never
a self-rating.

| Agent | Owns | Scored by |
|---|---|---|
| Interviewer | Vague brief → written spec; asks clarifying questions | User accepts the spec |
| Quant | Indicators, features, the maths | Rungs 2, 6 |
| Coder | Spec + maths → kernel or visualizer | Rungs 2–6 |
| Painter | The picture | Rung 7 |
| Fixer | Diagnostics → repair | The rung that failed |
| Reviewer | What verification cannot catch — look-ahead bias, unbounded buffers, wall-clock time | User acceptance, later reverts |

Painter is separate from Coder deliberately: composition and scale are a different skill from signal
logic, and splitting them means the draw path is scored on its own.

## Routing

From the AgentPack/ATR preprint, Derivation 1: `q*ₐ ∝ pₐ · e^(η·r̄ₐ)`, proven the unique trust-region
optimum.

**The whole simplification is that `p` is a prior and nothing requires it to come from a neural net.**
A hand-written state machine supplies it, so the same update rule works from trajectory zero and
inherits the Exp3 regret bound the paper's own remark notes. No training data, no model to ship.

```
new brief            → Interviewer
spec, needs maths    → Quant
maths ready          → Coder
compiles, no picture → Painter
any rung failed      → Fixer   (weighted by which rung)
all green            → Reviewer → done
```

Reliability is an EMA over verifier outcomes — closed form, no gradients, useful from the first run.

## Tasks

- [ ] `AgentRole`, and a `RoutingState` describing what the session knows
- [ ] `RoutingPrior`: the state machine, as a distribution
- [ ] `AgentReliability`: EMA, with a cold-start that is neither 0 nor 1
- [ ] `AgentRouter`: `q ∝ p·e^(η·r)`, deterministic pick plus seeded sampling for exploration
- [ ] Wire the ladder's outcome into reliability
- [ ] Agent prompts, one per role, composed onto the generated surface

## Constraints

**Reliability modulates the prior; it never overrides it.** An agent the state machine gives zero
probability must remain unreachable no matter how reliable it looks — Painter cannot be chosen before
there is code that compiles, and a bandit that could reach it would eventually try.

**The routing core takes no model.** It is a pure function of state, reliability and η, so it is
testable without a network, a key, or a token budget. Anything that needs a provider belongs in the
slice above it.

**Cold start is the normal case**, not an edge case. Most users will never accumulate enough
trajectories for the learning to matter, so the prior has to be right on its own and the bandit only
ever has to avoid making it worse.

**Comment by dhruuvsharma** [comment]
## Routing core landed — `1b21da4`

- [x] `AgentRole` and `RoutingState`
- [x] `RoutingPrior` — the state machine, as a distribution
- [x] `AgentReliability` — EMA with a neutral cold start
- [x] `AgentRouter` — `q ∝ p·e^(η·r)`, deterministic pick plus seeded sampling
- [x] `LadderFeedback` — the join between verifier and router
- [ ] Agent prompts, one per role, composed onto the generated surface
- [ ] Drive the loop from a real session

38 tests. Everything above is a pure function of state, reliability and η — no provider, no key, no
token budget — which is why it could be built before any agent exists to be routed to.

### Four decisions worth keeping

**A zero in the prior is absolute.** Reliability redistributes weight among agents the state machine
already allows; it can never make an ineligible one reachable. Painter before there is code that
compiles is not a worse choice, it is a meaningless one — and a bandit permitted to reach it would
eventually try.

**Cold start is the normal case, not an edge case.** Most users will never accumulate enough
trajectories for the learning to matter, so an unknown agent starts *neutral*: zero would multiply it
out of contention permanently, one would have every newcomer outrank proven peers.

**`Choose` is deterministic.** A user who re-runs a brief should get the same route; a session that
wanders cannot be debugged, and over an action space of six the exploration is worth far less than
that. Sampling exists, seeded, for when exploring is the point.

**Each rung routes to whoever is closest to it.** A draw-probe failure is the Painter's own work coming
back; a schema mismatch is the Coder having declared a parameter it then hard-coded. A general repair
agent loses the context that makes those cheap.

### The reward function had the reward hack in it

Worth recording plainly. `RewardFor` first returned `max(progress, 0.9)` for a passing report — so a
run that **skipped almost every rung** collected 0.9, identical to one that cleared them all.
Arranging to be checked by very little is precisely what an agent would have learned to do.

It is banded now: a pass scores 0.5–1.0 rising with how much was actually checked, a failure 0.0–0.5
rising with how far it got. Passing still dominates, because a passing artifact is deliverable — but
it no longer dominates for free. The denominator is fixed at eight rather than taken from the report,
or a run that stopped after one rung would score 1/1 and look perfect.

**My own tests caught it**, which is the argument for writing the adversarial half first.

### Next

Prompts per role, composed onto the generated SDK surface, and then driving the loop from a real
session so the reliability estimate starts seeing actual outcomes.

**Comment by dhruuvsharma** [comment]
## Phase 4's mechanism is complete — `d853af1`, `4b1e1e6`

- [x] `AgentRole`, `RoutingState`, `RoutingPrior`, `AgentReliability`, `AgentRouter`
- [x] `LadderFeedback` — the join between verifier and router
- [x] Six role instructions, composed as a separate system block
- [x] `AgentLoop` — route, ask, judge, record, repeat
- [ ] **Wire it to the authoring pane** — blocked on a product decision, below

58 tests. The whole thing runs from a scripted client, so a full multi-agent run can be asserted rather
than observed.

### The finding that changed the design

The system prompt was sent as **one cached block**:

```csharp
System: [new WireText(request.SystemContext) { CacheControl = Ephemeral }]
```

Append a role instruction to that and every agent becomes a different prefix — so **every role switch
re-bills the shared pack, roughly twelve thousand tokens, at full price.** Over a multi-turn session
with six agents the user buys that document a dozen times.

The role now travels as its own field and the client emits two blocks with the breakpoint on the
shared one. Six agents, one cached prefix, only a short role suffix ever re-read. The wire format
already took a list; only the code putting one thing in it had to change.

`StrategySkillLibrary` had recorded the same insight for skills — chosen once per session because the
system prompt is the cached prefix — and it applies here for exactly the same reason.

Fixing the CLI client turned up another four-file leftover: its closing instruction demanded *"kernel +
ITradingStrategy descriptor + live view-model + code-built view"*, so it had been telling a CLI agent
to write three files the host would never load.

### What the loop treats as normal rather than as failure

- **An agent answering without code.** For the Interviewer and Reviewer that *is* the job; for the
  others it means a question. The run pauses and waits.
- **A budget.** It is the user's money, and a loop that can spend without limit will. The honest end to
  an unsatisfiable brief is "here is what I built and what did not work".
- **The judge owning the state.** A report cannot know whether a unit owes a picture or whether a human
  reviewed it, so the judge returns the next state rather than the loop guessing.

Every turn keeps the posterior, not just the chosen agent — a trajectory of bare names is one nobody
can distil later, which is the whole point of logging it.

---

## Blocked: how much of this should a user see?

`AgentLoop` is deliberately **not** wired to the authoring pane, because the wiring encodes a product
decision about the user's time and tokens, and it should not be settled by one implementation being
the only thing that exists.

**A six-agent run costs roughly six times a single conversation.** Today the pane is one thread the
user watches and answers. Three plausible shapes:

1. **Silent** — the user gives a brief and meets the finished, verified unit. Fewest interruptions,
   most tokens spent before they can object.
2. **Visible** — each agent's turn appears in the transcript as it happens. Most legible, and the user
   can stop a run going wrong.
3. **Gated** — the user approves at the spec, and again before registration. Cheapest when a brief was
   misunderstood, most interrupting when it was not.

There is also a real question of whether the agent loop **replaces** the single-conversation builder or
sits beside it as a separate mode. The existing one works and is cheaper; making it the fallback rather
than deleting it is probably right, but that is a call about the product.

**Comment by dhruuvsharma** [comment]
## Constraint set: minimum tokens, extreme quality, time is free

That resolves the open question, and it changed the design rather than just picking an option.

### The loop was carrying a transcript

Every agent's reply appended to one list, re-sent every turn. Obvious design, wrong on both counts:

- **Cost is quadratic in turns.** Twenty turns of two-thousand-token replies means forty thousand
  tokens of history paid for on the last turn alone.
- **Quality drops with length.** It hands the Painter the Quant's derivation, the Fixer the
  Interviewer's questions, and everyone three stale copies of the same file. Models repair the version
  they can see; a stale one in context is a correctness failure before it is a cost one.

`AgentContext` replaces it with artifacts — brief, spec, maths, **current** code, **current** findings —
composed per role. Under this constraint the two goals stop competing: the cheapest context is also the
cleanest one.

**What each agent does not get is as deliberate as what it does.** The Painter gets no specification —
it must draw what the code computes, not re-derive intent. The Fixer gets none either: its findings
carry a remedy, and a Fixer that can see the original intent starts redesigning instead of repairing.
The Reviewer gets both, because comparing them is its whole job.

Two tests state the property rather than trusting the comment: twenty-five turns with the composed
context asserted not to grow, and the same twenty replies built both ways with the artifact form
required to come in under a quarter of the transcript.

### Shape: gated

The spec gate is the cheapest possible place to catch a misunderstood brief — one Interviewer turn
instead of six agents building the wrong thing. With time free, the interruption costs nothing that
matters. Registration stays gated too, since that is already how consent works today.

### The corollary worth stating

**Fewer, better turns beat more, cheaper turns — for tokens, not just for quality.** A failed turn pays
its full input and buys nothing, so under a token budget the effort setting should go *up*, not down.
Spending more to get a turn right the first time is the token-optimal move, and it only looks like a
trade-off if you count per-turn cost instead of per-delivery cost.

That also argues the agent loop should become the default rather than sit beside the single-conversation
builder: one conversation accumulates the same quadratic history, so for anything past a couple of
turns the artifact loop is now the cheaper of the two as well as the better one.

**Comment by dhruuvsharma** [comment]
## The pipeline runs end to end — `e733908`, `4d0e0bd`, `f4be92e`

- [x] Routing prior, reliability, router, ladder join
- [x] Six role instructions, sent as a separate cached-safe system block
- [x] `AgentContext` — artifacts, not a transcript
- [x] `AgentLoop`
- [x] `AuthoringJudge` — **all eight rungs, in one pass, for the first time**
- [x] Trajectory log (Phase 7, brought forward)
- [ ] Wire the gated flow into the pane — **needs a UI decision**

98 tests. A brief now goes: route → agent → extract → compile → eight rungs → reward → route again, all
provable against the real Roslyn compiler.

### The judge closed a gap nobody had noticed

Rungs 1–4 lived in the compiler and 5–8 in the verifier. **Both halves existed; nothing had joined
them**, so no code path had ever run the whole ladder. `AuthoringJudge` is the delegate `AgentLoop`
always took and nobody had written.

Three decisions in it: compile/policy/shape are *recorded* as cleared rather than omitted (the reward is
computed from that count, so omitting them silently underpays every agent that got that far);
`MustDraw` comes from the resolved type rather than the brief; and **`Reviewed` is a human's word, never
the ladder's** — a clean verdict that set it would route straight past the Reviewer, and the one agent
covering what verification cannot see would never run.

### Running it end to end immediately found a hole

A state with **code written, not yet compiled, and no failure recorded** fell through every branch of
the prior to an empty result — and an empty prior is how the loop is told to stop. It reported a unit
**delivered** having produced nothing.

Reading the state machine would not have shown that. Running a real compile through it did. The prior
now routes unjudged code to the Coder, and the empty case is guarded by `IsComplete` so it cannot be
reached from a state with work left in it.

### The log measures the thing being optimised

Per-turn tokens, and the **cached share** in particular — the figure the system-prompt split was for. If
it falls, something has broken the prefix, which costs money silently and shows up nowhere else.

It records numbers and finding codes, **never text**. Not the brief, not the reply, not the code. A
user's strategy is their property, and a log that swallowed model output would be a channel for feeding
untrusted text into a later prompt.

---

## Blocked: what the pane should look like

"Gated" settles *when* the user is asked. It does not settle what they see, and these are not
mechanical:

1. **Does the transcript show six agents, or one voice?** Six is honest and legible; it also means a
   user watching a brief go by sees the Interviewer, then the Quant, then the Coder, and may reasonably
   ask why they are paying for a committee.
2. **Where does the live preview sit in a multi-agent run?** It currently updates on compile. In a
   six-turn run that is six previews, most of them of half-finished work.
3. **Does the agent loop replace the single-conversation builder, or sit beside it as a mode?**

The mechanism is done and tested; none of this is blocked on more code. It is blocked on how you want
it to feel.

**Comment by dhruuvsharma** [comment]
## Decided: build effort is the selector

**Quick and Standard keep the single conversation. Deep and Max route through the agents.**

Effort already means what this decision is about — a user who picks Deep or Max has said correctness
over cost, and the agent split is precisely that trade. Reusing the dial means no second toggle to
explain, and **no way to ask for extreme quality and quietly get the cheap path**.

The decision lives in `StrategyBuildProfile`, which is already where an effort level's consequences are
written down, rather than in a new policy class.

### Two things that came with it

`MaxAgentTurns` is deliberately larger than `MaxFixAttempts` — a run spends turns on roles before it
spends any on repair, and interview → maths → code → picture → review is five before a single fix. A
budget that could not reach the Reviewer would spend the user's money and stop short of the one agent
covering what verification cannot see. Deep gets 10, Max gets 16.

`BacktestSmoke` on that record is now `Verify`. It had outlived the pass it named by a fortnight, and
the authoring pane was showing the user a task called **"Backtest smoke"** — which was simply untrue.

### The tests pin properties, not numbers

Effort never goes backwards; every agent run is verified; an agent budget can always reach the
Reviewer; Quick stays cheap; and an unrecognised value lands on Standard, because a configuration typo
must not silently start billing for six agents.

---

## Still open, and both are UI rather than mechanism

1. **Does the transcript show six agents or one voice?** My inclination is to show them, labelled,
   reusing the message stream that already exists — a six-turn run that looks like a long silence is
   worse than one that reads as work being done, and it justifies what it cost.
2. **Where does the live preview sit in a multi-agent run?** It updates on compile today, which in a
   six-turn run means several previews of half-finished work. That may be a feature rather than a
   problem — the preview already explains itself when there is nothing to show.

Neither is blocked on code. Both are how it should feel.

**Comment by dhruuvsharma** [comment]
## Phase 4 complete — wired, and closing

- [x] Routing prior, reliability, router, ladder join
- [x] Six role instructions, cached-prefix safe
- [x] `AgentContext` — artifacts, not a transcript
- [x] `AgentLoop`, `AuthoringJudge`, trajectory log
- [x] Build effort as the selector
- [x] **Wired to the authoring pane**

**Deep and Max run the six agents. Quick and Standard keep the single conversation.**

### Turns arrive as they happen

`AgentLoop` gained an `IProgress<AgentTurn>` and the pane appends each turn with the role as its
headline. A ten-turn run takes minutes; a pane showing nothing meanwhile reads as a hang. A user
watching six agents should be able to see **which one is spending their money and on what**.

The preview updates on every compile, half-finished work included — watching the picture arrive is
informative, and the preview already explains itself when there is nothing to show.

Reliability lives on the view-model, not per run: an agent's record is only worth having if it survives
the session that produced it.

### The budget message is honest rather than encouraging

A run that hits its ceiling says how far it got — whether it compiles — instead of inviting another
spend on the same wall. Under a token budget that is the entire point of having a ceiling.

---

## What the whole epic now does

A brief at Deep or Max: **route → agent → extract → compile → eight rungs → reward → route again**,
until the unit is delivered, an agent asks something, or the budget stops it. What comes out compiles,
has been driven through its real lifecycle, has drawn a frame that was inspected, has had its book read
back, previews live in the pane, and can be registered into the catalog and opened.

Every part of that was missing at the start of this work.

Closing. What remains of #44 is Phase 5's routing polish and whatever the first real run against a live
provider teaches — which is now a matter of using it rather than building it.
